osu.Game/Skinning/Skin.cs

```diff
@@ -27,9 +35,65 @@ public abstract class Skin : IDisposable, ISkin
 
         public abstract IBindable<TValue> GetConfig<TLookup, TValue>(TLookup lookup);
 
-        protected Skin(SkinInfo skin)
+        protected Skin(SkinInfo skin, IStorageResourceProvider resources)
         {
             SkinInfo = skin;
+
+            // we may want to move this to some kind of async operation in the future.
+            foreach (SkinnableTarget skinnableTarget in Enum.GetValues(typeof(SkinnableTarget)))
+            {
+                string filename = $"{skinnableTarget}.json";
+
+                // skininfo files may be null for default skin.
+                var fileInfo = SkinInfo.Files?.FirstOrDefault(f => f.Filename == filename);
+
+                if (fileInfo == null)
+                    continue;
+
+                var bytes = resources?.Files.Get(fileInfo.FileInfo.StoragePath);
+
+                if (bytes == null)
+                    continue;
+
+                string jsonContent = Encoding.UTF8.GetString(bytes);
+
+                DrawableComponentInfo[skinnableTarget] = JsonConvert.DeserializeObject<IEnumerable<SkinnableInfo>>(jsonContent).ToArray();
+            }
+        }
+
+        /// <summary>
+        /// Remove all stored customisations for the provided target.
+        /// </summary>
+        /// <param name="targetContainer">The target container to reset.</param>
+        public void ResetDrawableTarget(SkinnableElementTargetContainer targetContainer)
+        {
+            DrawableComponentInfo.Remove(targetContainer.Target);
+        }
+
+        /// <summary>
+        /// Update serialised information for the provided target.
+        /// </summary>
+        /// <param name="targetContainer">The target container to serialise to this skin.</param>
+        public void UpdateDrawableTarget(SkinnableElementTargetContainer targetContainer)
+        {
+            DrawableComponentInfo[targetContainer.Target] = targetContainer.CreateSkinnableInfo().ToArray();
+        }
```

Comment:
Could use the `ISkinnableTarget` interface instead of the container implementation itself.